src/sequence_number.rs

```diff
@@ -1,58 +1,163 @@
 use std::borrow::Borrow;
-use std::num::Wrapping;
+use std::convert::{TryFrom, TryInto};
 
-pub trait SequenceNumber {
-    type Rollup: Borrow<[u8]>;
-    fn rollup(&self) -> Self::Rollup;
-    fn root() -> Self
-    where
-        Self: Sized;
-    fn next_child(&mut self) -> Self
-    where
-        Self: Sized;
+pub trait UInt: TryFrom<usize> + Copy {
+    type Bytes: Borrow<[u8]>;
+    fn prime_init() -> Self;
+    fn prime_mult() -> Self;
+    fn to_le_bytes(self) -> Self::Bytes;
+    fn wrapping_add(self, other: Self) -> Self;
+    fn wrapping_mul(self, other: Self) -> Self;
 }
 
-#[derive(Debug)]
+pub trait SequenceNumber: Clone {
+    fn root() -> Self;
+    fn next_child(&mut self) -> Self;
+    fn skip(&mut self, count: usize) {
+        for _ in 0..count {
+            self.next_child();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct SequenceNumberInt<T> {
-    rollup: Wrapping<T>,
+    rollup: T,
     child: usize,
 }
 
-macro_rules! impl_sequence_no {
-    ($T:ty, $size:expr, $prime_init:expr, $prime_mult:expr) => {
-        impl Default for SequenceNumberInt<$T> {
-            fn default() -> Self {
-                Self::root()
-            }
+impl<T: UInt> Default for SequenceNumberInt<T> {
+    #[inline(always)]
+    fn default() -> Self {
+        Self::root()
+    }
+}
+
+impl<T: UInt> SequenceNumberInt<T> {
+    pub fn rollup(&self) -> T {
+        self.rollup
+    }
+}
+
+impl<T: UInt> SequenceNumber for SequenceNumberInt<T> {
+    fn root() -> Self {
+        Self {
+            rollup: T::prime_init(),
+            child: 0,
         }
-        impl SequenceNumber for SequenceNumberInt<$T> {
-            type Rollup = [u8; $size];
-            fn root() -> Self {
-                Self {
-                    rollup: Wrapping($prime_init),
-                    child: 0,
-                }
-            }
+    }
+    #[inline]
+    fn next_child(&mut self) -> Self {
+        let child = self.child;
+        self.child += 1;
 
-            #[inline]
-            fn next_child(&mut self) -> Self {
-                let child = self.child;
-                self.child += 1;
+        let rollup = self
+            .rollup
+            .wrapping_mul(T::prime_mult())
+            .wrapping_add(child.try_into().unwrap_or_else(|_| panic!("Overflow")));
```

Comment:
why `unwrap_or_else` instead of just `unwrap`, considering we are not adding any information in the message?

Reply:
`unwrap` would require that the trait `UInt`'s `TryFrom<usize>` have the bound that `TryFrom<usize, Error=::std::fmt::Debug>`. This isn't legal Rust today. So, the only way to get that trait bound would be to pollute a bunch of common method signatures with `where` clauses which I very much wish to avoid.

Reply:
I see.